### ISSUE TITLE
Performance getting worse afer wake up from sleeping repeatedly

### ISSUE THREAD

**Issue by jack338c** [5T33Z0/Thinkpad-T490-Hackintosh-OpenCore#44]
I don't know this is my t490 or your EFI problem.but everying is ok on windows 11
my t490 model is 20N3 too,this issue is still on sequoia 15.2 (24C101),I have to reboot ,otherwize pc is hot and runs slowly.
Could you tell me what will cause this issue probably?
I did't chang anyting,but use 16M itlwm.kext instead of yours
thanks

**Comment by 5T33Z0** [comment]
I think its a combination of multiple factors:

- The iGPU is spoofed as a different model since the Intel UHD 620 is not supported by macOS natively. And as soon as the iGPU is used it gets a lot hotter compared to how it behaves under Widows. When the system is getting hot it starts to throttle down the CPU frequency. I tried to adjust for with CPUFriend but it doesn't help much. I think the necessity to spoof the iGPU is the root cause of this.
- Another issue is that the air intake of the system is on the bottom of the system so when it sits on the desk not a lot of air can flow into it - and it gets worse once you actually put it on your lap

YogaSMC comes with a manual fan speed control accessible from the menubar. So you might want to ramp up the fan speed to max speed manually for a few minutes. I have tried various framebuffer patches and settings but I couldn't find a solution.

**Comment by jack338c** [comment]
it sound like I'm not alone.will try it later
Thank you very much for your detailed explain.

**Comment by jack338c** [comment]
BTW,I tested [this one](https://github.com/Trijal08/OpenCore-Hackintosh-Lenovo-ThinkPad-T490) these days,the performance is ok,but not fit for me.have trackpad tap to click issue,and crash problem.
thanks again for your great job 🥂

**Comment by 5T33Z0** [comment]
Try this dataprovider kext. It's low frequency setting drops the CPU down in idle to 400 Mhz. In my tests, this worked quite well.

[CPUFriendDataProvider.kext.zip](https://github.com/user-attachments/files/18134975/CPUFriendDataProvider.kext.zip)

**Comment by jack338c** [comment]
Many thanks
I disabled  CPUFriend, Dataprovider kext and YogaSMC as your mention and did't get the same issue until now.
also I tested the benchmark  before and now
now (disabled CPUFriend....) fan works ok too
![Screenshot 2024-12-14 at 16 26 51_nodate](https://github.com/user-attachments/assets/3c080bae-a50b-4508-8bba-bced312d67e0)
before
![Screenshot 2024-12-14 at 16 18 35](https://github.com/user-attachments/assets/56a91711-2e60-4d19-9056-d22ec6cc8571)

**Comment by 5T33Z0** [comment]
It has more performance now, because Minimum CPU clock higher when CPUFriend and teh DataProvider kexts are disabled this should result in even higher temps.

**Comment by Azezel300** [comment]
Hey, im getting really low results on geekbench, even without the yoga and i don't have cpufriend, but my score doesn't reach even 650 single, and just hits 2700 multi, temperature is not the problem but i don't know what could be, i have a t490 with the i5 8365U, can someone help me pls?

**Comment by 5T33Z0** [comment]
I can't verify your issue. CPU Score is fine and the CPU is reported correctly as well

![Image](https://github.com/user-attachments/assets/1a720ab1-e2e8-4974-8025-59fc077dff5c)

https://browser.geekbench.com/v6/cpu/11448146

![Image](https://github.com/user-attachments/assets/06e7e6f3-4b1c-47ad-ad57-ea798f1756cb)

**Comment by Azezel300** [comment]
Ok, thanks anyway, i will try to see what is the problem with mine cause i get almost half of your performance so maybe is just some config.

**Comment by Azezel300** [comment]
Hey, back to let the information if someone in the future has that problem, i fixed the slow speed by disabling the cpu friend, i think that because my i5 is a different model, it was capping his speed, now i think that if i make mt own cpufriend kext maybe that will work

Thx

**Comment by 5T33Z0** [comment]
I think the issue is Yoga SMC. I think the performance Slider in the prefpane doesnt work properly.

**Comment by Azezel300** [comment]
Ok, i will wait, for me the yoga doesn't make any diference in any performance mode, i made the benchmark as the result was around 650-720 on single and 2500-2700 on multi

When i disabled cpu provider and yogasmc, i got 1400 on single and 3900 on multi, but i didn't try disabling just yoga, i will make some changes to test, and i share with you

**Comment by 5T33Z0** [comment]
The new release is up.

**Comment by Azezel300** [comment]
> The new release is up.

Ok, i will test with my t490 and post the results after

**Comment by medkintos** [comment]
Hello 5T33Z0, the X1C7 user again here... got the same issue as @jack338c after several sleeps.

I think it is related to "Lap mode" that Lenovo implemented. If Lap Mode detected, the laptop's EC will throttle the CPU with BDPROCHOT ([you can read about it here](https://wiki.archlinux.org/title/Lenovo_ThinkPad_X1_Extreme#Disabling_BDPROCHOT_to_prevent_throttling_on_battery))

As YogaSMC didn't work on Sonoma+ on controlling the DYTC, we can't prevent the BDPROCHOT to kick in when it detects on lap mode...

But from the archlinux wiki mentioned above, we can make the "lap mode" disabled, by disabling BDPROCHOT entirely.

What i found for our hackintosh build that we can disable it by adding [DisablePROCHOT.efi](https://github.com/arter97/DisablePROCHOT) and [SimpleMSR.kext](https://github.com/arter97/SimpleMSR) for re-enabling the Disable PROCHOT protocol after S3 Sleep. So far works fine after several sleeps, but i need more time for testing.

Yes it will make the laptop runs hotter. But with the known YogaSMC issues, i think this is the better way to aid the performance of the laptop on macOS.

**Comment by 5T33Z0** [comment]
Wow, great finding. Thanks I will add the driver and kext to my EFI and test it.

Are there any terminal commands to check in what mode it is in, or do I need to re-enable YogaSMC and the required SSDTs?

**Comment by medkintos** [comment]
> Wow, great finding. Thanks I will add the driver and kext to my EFI and test it.
> 
> Are there any terminal commands to check in what mode it is in, or do I need to re-enable YogaSMC and the required SSDTs?

Yes, you must enable YogaSMC to know currently applied mode (as it read the EC.) 

Because I don't use YogaSMC anymore, i see it by eyeing the temp sensors (by app like Stats/HWSensors). If the temp >70C and still runs snappy, the BDPROCHOT bypass is working.

The right way to see it working is by eyeing the CPU frequency... CPU frequency will locked at 600-800Mhz if BDPROCHOT kicked in. If you know the app (besides Intel Power Gadget), tell me

After adding that, my single-core score is higher than before.... here's before and after on 3 runs on different range of time...

<img width="1787" height="1094" alt="Image" src="https://github.com/user-attachments/assets/15dd8cad-d4c2-4c5a-8dab-d07c5d221c4c" />

**Comment by 5T33Z0** [comment]
I added the driver and the kext to disable BDOROCHOT and reenabled YogaSMC. I noticed that after booting, the slider laways sits on "Quiet" (Lap). Only after moving it around the mode changes. And during the 2nd test, something strange happened when the multicore test kicked it. It Seems that the Mode fell back to Lap or something. See Screenshot:

<img width="1797" height="846" alt="Image" src="https://github.com/user-attachments/assets/cff0852f-520e-4f94-ac80-e736c90580f8" />

**Comment by 5T33Z0** [comment]
# Benchmarks

Tested wIth the added driver and kext to prevent BDPROCHOT.

## Test Evironment

**macOS**: 14.7.7 (23H709)
**CPU**: Intel Core i5 8265U (Quad Core)
**iGPU**: Intel UHD 620 (spoofed as UHD 630) 
**Geekbench**: 5.5.1

## CPU

Test # | Parameters | Single-Core <br>Score| Multicore <br>Score
:-----:|------------|:--------------------:|:--------------------:
[1](https://browser.geekbench.com/v5/cpu/23665259) | <ul><li>YogaSMC **OFF** <li> CPUFriend: **OFF** |1017|3252
[2](https://browser.geekbench.com/v5/cpu/23665246) |<ul> <li>YogaSMC **OFF** <li> CPUFriend: **ON** | 1016 | 3330

## iGPU

Setup # | Parameters | Freq. (GHZ) | OpenCL<br>Scrore | Metal<br>Score 
:-----:|-------------|:-----------:|:----------------:|:--------------:
1 | <ul> <li>CPUFriend: **OFF** <li>[GUC FW](https://github.com/5T33Z0/OC-Little-Translated/blob/main/Content/11_Graphics/iGPU/GUC_Firmware.md): **ON**| 1,10| [4927](https://browser.geekbench.com/v5/compute/6878344) | [4660](https://browser.geekbench.com/v5/compute/6878345)
2 | <ul> <li> CPUFriend: **OFF** <li>[RPS Control](https://github.com/5T33Z0/OC-Little-Translated/blob/main/Content/11_Graphics/iGPU/RPS-Control.md): **ON** | 110 | [**4930**](https://browser.geekbench.com/v5/compute/6878349) |[4700](https://browser.geekbench.com/v5/compute/6878350)
3 | CPUFriend: **OFF** | 1,10 | [4559](https://browser.geekbench.com/v5/compute/6878351) | [4418](https://browser.geekbench.com/v5/compute/6878352)
4 | <ul> <li>CPUFriend: **ON** <li>GUC FW: **ON** | 1,10| [4858](https://browser.geekbench.com/v5/compute/6878353) | [4620](https://browser.geekbench.com/v5/compute/6878354)
5 | <ul> <li>CPUFriend: **ON** <li>RPS-Control: **ON** | 1,10 | [4841](https://browser.geekbench.com/v5/compute/6878355) | [**4718**](https://browser.geekbench.com/v5/compute/6878356)
6 | CPUFriend: **ON** | 1,10 | [4684](https://browser.geekbench.com/v5/compute/6878361)| [4411](https://browser.geekbench.com/v5/compute/6878363)

**Comment by 5T33Z0** [comment]
I just searched for BDPROCHOT in gitub and found this kext: 

https://github.com/black-dragon74/NoBdProcHot

I wonder if this would work as a replacement for the efi and simplemsr.kext

**Comment by medkintos** [comment]
Update:

The CPU now throttles when the GPU is working hard.... (e.g Browser GPU Renderer). When i close my browser (I use Arc), the systems performance back to normal (see below)

<img width="512" height="880" alt="Image" src="https://github.com/user-attachments/assets/c2c17a4b-fc87-4d4b-ad23-563040e58b6d" />

If you look the red box, when GPU freqs is high, the CPU throttled to 700-800mhz. The GPU usage spiked because of the browser's GPU helper. After I closed the Arc browser, the GPU frequency went down, the CPU stopped throttling, and the system became responsive again.

In the past, once it started throttling, the system would remain laggy even after I closed all applications. I must restart the notebook to fix the performance. Now, it's running smoothly... It seems the BDPROCHOT issue somehow has been resolved. I will conduct further testing over the next few days before my honeymoon...

**Comment by jack338c** [comment]
I want to test it too.how to add these stuff @5T33Z0 
Currently myt490 efi is ThinkPad T490 OpenCore 1.0.5 v5.0

**Comment by 5T33Z0** [comment]
Maybe you can try this newer fork of VoltageShift instead: https://github.com/asepms92/VoltageShift

**Comment by medkintos** [comment]
Yes it works now using VoltageShift. I use Hamerspoon to disables BD Procohot everytime it wakes from sleep... Thankyou by the way! The launchd builder sadly doesn't provide options to disable prochot after wake...

I thought undervolting can be applied but it cant because OC_Locked by Lenovo itself...

As added bonus, using VoltageShift we can confirm that BD Prochot is disabled by read the MSR bit directly just like on linux

<img width="591" height="357" alt="Image" src="https://github.com/user-attachments/assets/7206a1db-dce6-4886-959d-993dfd439440" />

But if SimpleMSR works on your notebook, i suggest using that as it reduce dependencies (like using Hammerspoon mentioned above just to disable the prochot)

**Comment by jack338c** [comment]
Thanks again @medkintos 
The hi score one is from wake up
and low one is from reboot

<img width="1156" height="487" alt="Image" src="https://github.com/user-attachments/assets/5c0b9876-d34d-4299-9204-5d50bc520296" />
<img width="1212" height="510" alt="Image" src="https://github.com/user-attachments/assets/7dd88bc0-d784-4106-933e-368a345199e7" />

**Comment by 5T33Z0** [comment]
@medkintos Congrats. Can you upload your build of the volageshift kext so I can try it? I am currently on the road with limited acces to wifi and don't have Xcode installed so I can't compile it. Thanks

**Comment by medkintos** [comment]
tbh I still use sicreative's VoltageShift build because i can't build it... and then i realized that the clonned repo MacKernelSDK folder is empty... dumb me

[here is it](https://github.com/user-attachments/files/21306600/VoltageShift.zip)

**Comment by 5T33Z0** [comment]
Thanks. For me, results are worse than using smiplemsr.kext and CPUFriend. I think, I need an SSDT to control fan speed, if the CPU activity increases. YogaSMC did hande this quite well, but without it, the fan just stays at a constant speed once it turns on. So annoying.

**Comment by medkintos** [comment]
@5T33Z0 Looks like I've been doing the prochot disable wrong this whole time. It's not supposed to be set to 0...

The right way to disable Prochot is MSR 0x1FC values when BD Prochot is enabled, minus one bit [here's the article](https://github.com/yyearth/turnoff-BD-PROCHOT)

e.g. if your MSR value when BD Prochot enabled is 0x14001b, write 0x14001a to disable it.

So what i do is:

1.  Use DisablePROCHOT.efi to disable Prochot on first boot
2. Read the values with VoltageShift by `./voltageshift read 0x1fc`
3. Convert the result, my result is `00000000 00000000 00000000 00000000 00000000 00010100 00000000 00011010` into hex --> `0x14001a`
4. Use this value to write the MSR 0x1fc everytime it resets after sleep (I automated it with Hamerspoon) `./voltageshift write 0x1FC 0x14001a`

If you don't want to use DisablePROCHOT.efi:

1. Read the values with VoltageShift by `./voltageshift read 0x1fc`
2. Subtract the last bit of your result. e.g. if the result is `00000000 00000000 00000000 00000000 00000000 00010100 00000000 00011011` subtract it and you get `00000000 00000000 00000000 00000000 00000000 00010100 00000000 00011010`
3. Convert the subtracted result into hex --> `0x14001a`
4. Use this value to write the MSR 0x1fc everytime it resets after sleep (I automated it with Hamerspoon) `./voltageshift write 0x1FC 0x14001a`

And i just realized when you said about fan speed... my fan doesn't ramp up as high with YogaSMC... So i try to bench with YogaSMC + CPUFriend + right VoltageShift usage and surprised that my score is higher than before:

<img width="902" height="1051" alt="Image" src="https://github.com/user-attachments/assets/465a2bad-9f70-4214-959f-be12c43268b4" />

Will test the long term usage soon... gonna prepare for my wedding tomorrow... I'll probably check it out again once I'm back from the honeymoon.

---

Before i realize what i do was wrong (Disabling Prochot by zero-ing the MSR bit by `./voltageshift write 0x1fc 0`) i took the benchmark and got this low score (22 minutes before benchmark result above):

<img width="914" height="1071" alt="Image" src="https://github.com/user-attachments/assets/1f23a38c-e1f8-4b71-abc4-4477941927d1" />

**Comment by 5T33Z0** [comment]
Hmm, so I added the voltageshift.kext, rebooted and ran the voltageshift utility via terminal and entered: `./voltageshift read 0x1fc`

This returns all 0s and it does not stop:

<img width="1495" height="553" alt="Image" src="https://github.com/user-attachments/assets/f7f2d4e7-1127-4088-8815-e8724151cb1a" />

**Comment by medkintos** [comment]
@5T33Z0 sorry, i forgot to tell you to use the sicreative's one (1.2.5)....

**Comment by 5T33Z0** [comment]
I created an article about disabling BDPROCHIT in my OC-Little Translated Repo to cover the topicin general.

https://github.com/5T33Z0/OC-Little-Translated/tree/main/Content/05_Laptop-specific_Patches/BDPROCHOT 

I think, I can close this Issue now since it's resolved.

**Comment by 5T33Z0** [comment]
Looks to me like this is for AMD Laptops only: https://www.youtube.com/watch?v=qrIuEIlP4Xw

**Comment by medkintos** [comment]
there's possibility to use smokeless_umaf on Intel based machines: https://github.com/DavidS95/Smokeless_UMAF/discussions/6

someone tried it on Intel machine, but can't write changes https://www.reddit.com/r/thinkpad/comments/1gfdd1z/anyone_tried_umaf_on_a_t495/

**Comment by jack338c** [comment]
@5T33Z0 
Thanks for your good job.my t490 runs more smoothly now !
Currently my t490 got some little issue with the Lenovo ThinkPad T490 OpenCore 1.0.6 v1.0 
1 tap with two fingers response is not sensitive,but tap three fingers is ok.
2 trackpad and trackpoint will be disabled from sleep to wake sometime.
3 play full screen hd video will be frozen on app APTV(iptv app) ,exit full screen,issue gone.
Hope it doesn't just happen on my t490.
